taskcat/testing/manager.py

```diff
@@ -0,0 +1,252 @@
+                buckets, self.config.config.project.name, self.config.project_root
+            )
+
+        regions = self.config.get_regions(boto3_cache)
+        parameters = self.config.get_rendered_parameters(buckets, regions, templates)
+        tests = self.config.get_tests(templates, regions, buckets, parameters)
+
+        self.test_definition = Stacker(
+            self.config.config.project.name,
+            tests,
+            shorten_stack_name=self.config.config.project.shorten_stack_name,
+        )
+        self.test_definition.create_stacks()
+        terminal_printer = TerminalPrinter(minimalist=self.minimal_output)
+
+        terminal_printer.report_test_progress(stacker=self.test_definition)
+
+    def end(
+        self,
+        no_delete: bool = False,
+        keep_failed: bool = False,
+        dont_wait_for_delete: bool = True,
+    ):
+        """
+        docstring
+        """
+
+        terminal_printer = TerminalPrinter(minimalist=self.minimal_output)
+
+        status = self.test_definition.status()
+
+        # Delete Stacks
+        if no_delete:
+            LOG.info("Skipping delete due to cli argument")
+        elif keep_failed:
+            if len(status["COMPLETE"]) > 0:
+                LOG.info("deleting successful stacks")
+                self.test_definition.delete_stacks({"status": "CREATE_COMPLETE"})
+        else:
+            self.test_definition.delete_stacks()
+
+        if not dont_wait_for_delete:
+            terminal_printer.report_test_progress(stacker=self.test_definition)
+
+        # TODO: summarise stack statusses (did they complete/delete ok) and print any
+        #  error events
+
+        # Delete Templates and Buckets
+        buckets = self.config.get_buckets()
+
+        if not no_delete or (keep_failed is True and len(status["FAILED"]) == 0):
+            deleted: ListType[str] = []
+            for test in buckets.values():
+                for bucket in test.values():
+                    if (bucket.name not in deleted) and not bucket.regional_buckets:
+                        bucket.delete(delete_objects=True)
+                        deleted.append(bucket.name)
+
+        # 9. raise if something failed
+        # - grabbing the status again to ensure everything deleted OK.
+
+        status = self.test_definition.status()
+        if len(status["FAILED"]) > 0:
+            raise TaskCatException(
+                f'One or more stacks failed tests: {status["FAILED"]}'
+            )
+
+    def report(
+        self, output_directory: str = "./taskcat_outputs",
+    ):
+        """
+        Generates a report of the status of Cloudformation stacks.
+        """
+        report_path = Path(output_directory).resolve()
+        report_path.mkdir(exist_ok=True)
+        cfn_logs = _CfnLogTools()
+        cfn_logs.createcfnlogs(self.test_definition, report_path)
+        ReportBuilder(
+            self.test_definition, report_path / "index.html"
+        ).generate_report()
```

Comment:
I made report its own little method instead of leaving this logic in the run command. Thought it might come in useful even when importing taskcat.

Reply:
In the future, I think report should be in the abstract base printer class, see my comments (rants) further down.